src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -325,6 +329,91 @@ EnsureCitusTableCanBeCreated(Oid relationOid)
 }
 
 
+/*
+ * undistribute_table gets a distributed table name and
+ * udistributes the table.
+ */
+Datum
+undistribute_table(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+	EnsureCoordinator();
+
+	Oid relationId = PG_GETARG_OID(0);
+	const char *relationName = get_rel_name(relationId);
+
+	EnsureTableOwner(relationId);
+
+	Relation relation = relation_open(relationId, ExclusiveLock);
+
+	if (TableReferencing(relationId))
+	{
+		ereport(ERROR, (errmsg("Cannot undistribute table "
+							   "because it has a foreign key.")));
+	}
+
+	if (TableReferenced(relationId))
+	{
+		ereport(ERROR, (errmsg("Cannot undistribute table "
+							   "because a foreign key references to it.")));
+	}
+
+	StringInfo query = makeStringInfo();
+
+	SPI_connect();
+
+	List *tableDDLEvents = GetTableDDLEvents(relationId, true);
+	char *createTableQuery = linitial(tableDDLEvents);
+	list_delete_first(tableDDLEvents);
+
+	Node *parseTree = ParseTreeNode(createTableQuery);
+
+	((CreateStmt *) parseTree)->relation->relname = "tmp1";
+
+	CitusProcessUtility(parseTree, createTableQuery, PROCESS_UTILITY_TOPLEVEL, NULL,
+						None_Receiver, NULL);
+
+	resetStringInfo(query);
+	appendStringInfo(query, "INSERT INTO tmp1 SELECT * FROM %s", relationName);
+	SPI_execute(query->data, false, 0);
+
+	ObjectAddress undistributeTableAddres;
+
+	undistributeTableAddres.classId = RelationRelationId;
+	undistributeTableAddres.objectId = relationId;
+	undistributeTableAddres.objectSubId = 0;
+
+	DirectFunctionCall3(master_drop_all_shards,
```

Comment:
Could we just call the underlying C function directly / wrap the relevant bits of master_drop_all_shards in a C function that we can call directly?